### PR TITLE
Feature/characters

### PR DESCRIPTION
**Pull request by smartscripts-nl**
With this plugin it is possible to use selected text in an ebook for describing characters in the ebook:

1) When you want to describe a character, select his or her name. Activate the plugin (through a button defined in readerhighlight.lua, see below). An edit dialog with the selected text will be opened. Add a description to the name and save the character.

2) Later on text selections which contain (part of) the name of the character can be used to lookup the character, via the same route. When part of the name is found in the table with characters, the description will be shown, in stead of opening an edit dialog.

Handy for ebooks with lots of characters in them.

The plugin uses a helper widget, placed in ui/extensions. Maybe not the best of places, can of course be put somewhere else.

To use this plugin in my personal KOReader I've adapted readerhighlight.lua, to implement a button for defining/looking up characters:

{
                text = 'Character',
                callback = function()

                    self:highlightFromHoldPos()

                    if self.selected_text then

                        self.ui:handleEvent(Event:new("SaveCharacter", self.selected_text.text))

                        self:onClearHighlight()

                        UIManager:close(self.highlight_dialog)
                    end
                end,
            },

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6255)

### CODE REVIEW

**Comment by poire-z** [comment]
> The plugin uses a helper widget, placed in ui/extensions. Maybe not the best of places, can of course be put somewhere else.

Really not fond of that :/
You're introducing a 2nd API (that mostly just wrap with zero or only a few added lines our functions and UI, and Lua function/string and table methods), like if you were not happy with how Lua and us present stuff, and thought: I'll rewrite the function calls and names in my own words and way of thinking :)

_You_ should speak our language :) - and we should not encourage people to learn another - and we should not have to maintain a duplicated API. If you think some function and utils are missing, you should add them to our relevant places and modules, so people can find them where they would expect it - and they don't go pick a bit in our core, and a bit (the same, with another name :) in your extensions/helpers.lua...

As for the plugin, I'm not a customer for that - but I feel it might be too specific.
A more generic plugin would allow recording/remembering any "topics", "words", "references", "bibliography", which would allow it to be used in other contexts/workflows.
Or tomorrow, somebody will add "notions.plugin", "references.plugin", "years.plugin", "clues.plugin" (useful in crime novels! :)...

File: frontend/extensions/helpers.lua [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+		if current_ebook ~= nil then
+			homedir = string.gsub(current_ebook, '/[- .,:;?!_A-Za-z0-9]+', '')
+		end
+    end
+
+	return homedir
+end
+
+function Helpers:getKoreaderDir()
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:48]: I agree with @poire-z that there's a lot of stuff in this file that doesn't really seem to fit. This one for example is doing the reverse of `DataStorage:getSettingsDir()`, which is just `DataStorage:getDataDir()` + `/settings`/ I believe?

Comment (smartscripts-nl) [reply frontend/extensions/helpers.lua:48]: Maybe you're right. I'll look it up. I've learned lua by looking at the KOReader source code, so my code will not be always optimal.

**Comment by smartscripts-nl** [comment]
OK, just testing the water. I've also written a plugin which saves texts for inspirational quotes, which can be viewed and, if one wishes, displayed automatically every quarter hour. But from your reactions I gather you won't be enthusiastic about that one too.

I've written many other things, for example an improvement (imho) of the file search script. Maybe I'll test the water some more in the time to come.

**Comment by smartscripts-nl** [comment]
NB: in the code you can see that I sometimes repaint a dialog, when showing it. I don't know if my method is the best method to go about that, but I needed this behaviour to fix a problem on my Kobo Forma: the borders of dialogs are not always painted completely, parts of them are often missing. When I repaint the same dialog, the problem is gone.

**Comment by poire-z** [comment]
> the borders of dialogs are not always painted completely, parts of them are often missing. When I repaint the same dialog, the problem is gone.

Yeah, there's quite a bit of nastiness here and there :)
https://github.com/smartscripts-nl/koreader/blob/a8d270a15da3cdadb278cc0690f06e6f8c2a0af5/plugins/characters.koplugin/main.lua#L328-L342
(One minor thing that makes it hard to get into your code is all the blank lines.)

> I've learned lua by looking at the KOReader source code

Like most of us did :)
But there's Lua to learn, and there's KOReader's way of doing things (which is harder than Lua). You usually learn that by copying stuff from things that work, getting to understand the magic, and adapting it.
So, you picked the `UIManager:show()` which looked simple enough - but there's the whole business on refresh (that _is_ complicated, more complicated than other UI framework, because we target eInk panels) that you didn't pick: the `setDirty()` that is often set after showing stuff, to indicate the eInk driver which region to repaint. I guess that's something like that that is missing.
Your show/close/show double repaint solves that, I guess, by the fact that when closing, you're having the lower widgets asked to repaint themselves, some of them take the fullscreen, so it does `setDirty(the whole screen)`, then you show your little widget - and our UIManager enqueue/merge the setDirty regions to do a single refresh at the end (because eInk is slow) - so your little widget that you show() a 2nd time before the refresh is asked benefit from the fullscreen refresh - and is painted fully.

**Comment by smartscripts-nl** [comment]
> > The plugin uses a helper widget, placed in ui/extensions. Maybe not the best of places, can of course be put somewhere else.
> 
> Really not fond of that :/
> You're introducing a 2nd API (that mostly just wrap with zero or only a few added lines our functions and UI, and Lua function/string and table methods), like if you were not happy with how Lua and us present stuff, and thought: I'll rewrite the function calls and names in my own words and way of thinking :)
> 
> _You_ should speak our language :) - and we should not encourage people to learn another - and we should not have to maintain a duplicated API. If you think some function and utils are missing, you should add them to our relevant places and modules, so people can find them where they would expect it - and they don't go pick a bit in our core, and a bit (the same, with another name :) in your extensions/helpers.lua...
> 
> As for the plugin, I'm not a customer for that - but I feel it might be too specific.
> A more generic plugin would allow recording/remembering any "topics", "words", "references", "bibliography", which would allow it to be used in other contexts/workflows.
> Or tomorrow, somebody will add "notions.plugin", "references.plugin", "years.plugin", "clues.plugin" (useful in crime novels! :)...

OK, I was just testing the waters. It seems I don't fit in with your team. But keep up the good works!

**Comment by Frenzie** [comment]
Don't give up, I'll try to give some feedback in a more concrete manner. ;-)

File: frontend/extensions/helpers.lua [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+end
+
+function Helpers:shuffle(tbl)
+	for i = #tbl, 2, -1 do
+		local j = math.random(i)
+		tbl[i], tbl[j] = tbl[j], tbl[i]
+	end
+end
+
+function Helpers:split(str, pat)
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:111]: This seems to be used solely as a helper function for split_path? For split in general, see http://koreader.rocks/doc/modules/util.html#gsplit

File: frontend/extensions/helpers.lua [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+		s, e, cap = str:find(fpat, last_end)
+	end
+	if last_end <= #str then
+		cap = str:sub(last_end)
+		table.insert(t, cap)
+	end
+	return t
+end
+
+function Helpers:split_path(str)
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:130]: Not quite the same thing as http://koreader.rocks/doc/modules/util.html#splitFilePathName but without seeing what it's used for it's hard to say one way or the other.

File: frontend/extensions/helpers.lua [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+end
+
+function Helpers:ucfirst(str)
+	return (str:gsub("^%l", string.upper))
+end
+
+function Helpers:getMiddleDialogWidth()
+
+	local orientation = Screen:getScreenMode()
+	local iwidth = (Screen:getWidth() / 2) + 20
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:155]: More of a side note, but you'll generally want to wrap any numbers like that in `Screen:scaleBySize()`.

File: frontend/extensions/helpers.lua [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+	local iwidth = Screen:getWidth() - 120
+
+	if orientation == 'portrait' then
+		iwidth = Screen:getWidth() - 80
+	end
+
+	return iwidth
+end
+
+function Helpers:round(x)
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:176]: We actually have a bunch of math helpers like that in optmath.

I'll add some comment strings in so they'll show up in the docs.

File: frontend/extensions/helpers.lua [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+	if x % 2 ~= 0.5 then
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:178]: As an aside, this is not how I round. 1.5 rounds to 2, not to 1.

File: frontend/extensions/helpers.lua [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+	end
+
+	return x - 0.5
+end
+
+function Helpers:substrCount(subject, needle)
+	return select(2, subject:gsub(needle, ""))
+end
+
+function Helpers:readSetting(options_file, setting, default)
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:189]: At first glance this looks like needless duplication, but if it's not it should probably be fixed in LuaSettings.

File: frontend/extensions/helpers.lua [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+function Helpers:removeListItemNumber(text)
+
+	return string.gsub(text, '^[0-9]+[.][ ]*', '')
+end
+
+function Helpers:confirm(question, callback)
+
+	UIManager:show(ConfirmBox:new {
+		text = question,
+		ok_text = _("Yes"),
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:234]: Note that you can just leave those empty for the default values.

File: frontend/extensions/helpers.lua [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+		cancel_text = _("No"),
+		ok_callback = function()
+			callback()
+		end,
+	})
+end
+
+function Helpers:exists(path)
```
Comment (Frenzie) [inline frontend/extensions/helpers.lua:242]: http://koreader.rocks/doc/modules/util.html#pathExists

File: plugins/characters.koplugin/main.lua [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+
+        self.filter = character
+
+        return {
+            character = item_with_most_hits['character'],
+            characterno = item_with_most_hits['characterno'],
+        }
+    end
+
+    return nil
```
Comment (Frenzie) [inline plugins/characters.koplugin/main.lua:215]: Not very important, but you can just leave that out.

**Comment by smartscripts-nl** [comment]
The problem is maybe that I am a PHP programmer and I think Lua is sometimes rather clunky. So I like to write wrapper methods which to my taste (biased by PHP) are easier to "read" and to use.

The methods in helpers.lua can and should of course be placed somewhere else (if handy at all), but for my personal development work it is easier to keep them all together in one place.

File: frontend/extensions/helpers.lua [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+	else
+		UIManager:show(InfoMessage:new { text = message, icon_file = 'resources/info-bug.png', timeout = timeout, dismiss_callback = dismiss_callback })
+	end
+end
+
+function Helpers:alertDebugOnce(message, timeout, dismiss_callback)
+
+	if not self.debug_once_used then
+
+		self:alertDebug(message, timeout, dismiss_callback)
```
[the remark this answers is not in the document]
Comment (smartscripts-nl) [reply frontend/extensions/helpers.lua:70]: When I want to view values or until how far a script proceeds without errors, on my computer I insert temporary calls to alertDebug of alertDebugOnce, so I can see onscreen in KOReader itself how things work out. Completely developing on the Kobo itself would be very impractical.

**Comment by smartscripts-nl** [comment]
> One minor thing that makes it hard to get into your code is all the blank lines

Funny, to my taste the blank lines make the code more readable, yield better oversight... One can't argue about taste, but should I contribute code in the future, then I'll remove the blank lines.

File: frontend/extensions/helpers.lua (hunk 6 above)
Comment (smartscripts-nl) [reply frontend/extensions/helpers.lua:178]: When you run the code, you'll see that it rounds 1.5 nicely to 2:
 1.5 % 2 = 0 => so not 0.5, so return math.floor(1.5 + 0.5) => 2

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: 1.5 mod 2 = 1.5, so you return 1.5 - 0.5 = 1

0.7 mod 2 = 0.7, so you return 0.7 - 0.5 = 0.2

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: Wait, you and/or I confused me, but this is how your function behaves pasted into a Lua console. >_>

```
> print (Helpers:round(1.5))
1
> print (Helpers:round(2))
1.5
> print (Helpers:round(2.5))
2
> print (Helpers:round(0.7))
0.2
> print (Helpers:round(2.5))
2
> print (Helpers:round(1.5))
1
```

In any case the exception makes no sense.

```
function round(x)

	if x % 2 ~= 0.5 then
		return math.floor(x + 0.5)
	end

	return x - 0.5
end

print(round(.5))
print(round(1.5))
```
```
$ lua bla.lua 
0
2
```

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: Don't we already have a round in optmath?

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: (Which should essentially be doing floor + 0.5).

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: We do. GH may have helpfully hidden my comment saying exactly that right above. ;-)

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: Why, yes we do, and yes it does ;).

https://github.com/koreader/koreader/blob/23862fd0ebd51f8a21ed4dd82861806f356f6128/frontend/optmath.lua#L43

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: This one: https://github.com/koreader/koreader/pull/6255#discussion_r438648691

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: Oh, actually, it might be wrong. Does floor(n + 0.5) == trunc(floor(n) + 0.5)?

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: I don't think trunc is a thing in Lua, but that's probably the same thing. Why is that wrong?

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: I dunno, but in C I'd have done floor(n) + 0.5f cast to an integer (so, truncated).

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: My bad, I was assuming trunc(n + 0.5) as trunc(floor(n) + 0.5) looks completely meaningless to me.

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: As in:

floor(1.9) = 1
trunc(1.9) = 1
trunc(1 + 0.5) = 1

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: It'll be different for negative integers of course.

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: Yeah, no, I still suck at maths ;D.

floor(n) + 0.5f essentially does a ceil, not a round.

TL;DR: Everything's fine :D.

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: Never mind me, I still suck at maths ;D.

(i.e., floor(n + 0.5) works just fine ;)).

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: And I was actually thinking of `(int) (n + 0.5f)`, and not the nonsense I mentioned earlier ^^.

Comment (NiLuJe) [reply frontend/extensions/helpers.lua:178]: (So, basically, yeah, floor() :D).

Comment (Frenzie) [reply frontend/extensions/helpers.lua:178]: Difference illustrated in JS (roundToZero would be a much clearer name, cf. our [roundAwayFromZero](http://koreader.rocks/doc/modules/optmath.html#Math.roundAwayFromZero)):
```
Math.floor(-1.9)
-2
Math.trunc(-1.9)
-1
```
